### PR TITLE
Add support for resource `harbor_config_security`

### PR DESCRIPTION
**Pull request by naitmare01**
This PR adds support for the resource `harbor_config_security` which should fix the issue https://github.com/goharbor/terraform-provider-harbor/issues/223. 

Please note that at the moment Harbor doesn't have any apis to remove the system-wide cve allow list.

### CODE REVIEW

**Comment by Vad1mo** [comment]
thank you for your contribution, @naitmare01 do you have an idea how we can have an automated test setup?

**Comment by naitmare01** [comment]
We can probably write a test that performs all the supported operations on the list. E.g. update the list with strings of our choice. The textbox that stores all the cve-ids accept any strings really.

File: provider/resource_config_security.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+		Delete: resourceConfigSecurityDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+	}
+}
+
+func resourceConfigSecurityCreate(d *schema.ResourceData, m interface{}) error {
+	apiClient := m.(*client.Client)
+	body := SystemCVEAllowListBody(d)
```
Comment (flbla) [inline provider/resource_config_security.go:51]:
```suggestion
	body := client.SystemCVEAllowListBody(d)
```

File: provider/resource_config_security.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+func resourceConfigSecurityRead(d *schema.ResourceData, m interface{}) error {
+	apiClient := m.(*client.Client)
+
+	resp, _, respCode, err := apiClient.SendRequest("GET", models.PathSystemCVEAllowList, nil, 200)
+	if respCode == 404 && err != nil {
+		d.SetId("")
+		return fmt.Errorf("resource not found %s", d.Id())
+	}
+
+	var jsonData SystemCveAllowListBodyPost
```
Comment (flbla) [inline provider/resource_config_security.go:70]:
```suggestion
	var jsonData models.SystemCveAllowListBodyPost
```

File: provider/resource_config_security.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	d.Set("project_id", 0) //Setting this to zero to be able to force a system allow list since we are setting project lists in `harbor_project`
+	d.Set("update_time", jsonData.UpdateTime)
+	d.Set("creation_time", jsonData.CreationTime)
+
+	return nil
+}
+
+func resourceConfigSecurityUpdate(d *schema.ResourceData, m interface{}) error {
+	apiClient := m.(*client.Client)
+	body := SystemCVEAllowListBody(d)
```
Comment (flbla) [inline provider/resource_config_security.go:95]:
```suggestion
	body := client.SystemCVEAllowListBody(d)
```

File: provider/resource_config_security.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+	if err != nil {
+		return err
+	}
+
+	return resourceConfigSecurityRead(d, m)
+}
+
+func resourceConfigSecurityDelete(d *schema.ResourceData, m interface{}) error { // Harbor doesn't really have aby way to delete this resource yet...
+	apiClient := m.(*client.Client)
+	body := SystemCVEAllowListBody(d)
```
Comment (flbla) [inline provider/resource_config_security.go:107]:
```suggestion
	body := client.SystemCVEAllowListBody(d)
```

**Review by flbla — CHANGES_REQUESTED** [review]

File: provider/resource_config_security.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	return resourceConfigSecurityRead(d, m)
+}
+
+func resourceConfigSecurityRead(d *schema.ResourceData, m interface{}) error {
+	apiClient := m.(*client.Client)
+
+	resp, _, respCode, err := apiClient.SendRequest("GET", models.PathSystemCVEAllowList, nil, 200)
+	if respCode == 404 && err != nil {
+		d.SetId("")
+		return fmt.Errorf("resource not found %s", d.Id())
```
Comment (flbla) [inline provider/resource_config_security.go:67]:
```suggestion
		return fmt.Errorf("resource not found %s", models.PathSystemCVEAllowList)
```

File: provider/resource_config_security.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+	}
+
+	var jsonData SystemCveAllowListBodyPost
+	err = json.Unmarshal([]byte(resp), &jsonData)
+	if err != nil {
+		return fmt.Errorf("resource not found %s", d.Id())
+	}
+
+	id := jsonData.ID
+	stringid := fmt.Sprintf("%v", id)
```
Comment (flbla) [inline provider/resource_config_security.go:77]:
```suggestion
```

File: provider/resource_config_security.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+	expires_at, expires_at_true := d.GetOk("expires_at")
+
+	if expires_at_true {
+		d.Set("expires_at", expires_at)
+	}
+
+	d.SetId(stringid)
```
Comment (flbla) [inline provider/resource_config_security.go:85]:
```suggestion
	d.SetId(models.PathSystemCVEAllowList)
```

File: models/systemcve.go [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package models
+
+type SystemCveAllowListBodyPost struct {
+	ID           int                     `json:"id,omitempty"`
```
Comment (flbla) [inline models/systemcve.go:4]:
```suggestion
	ID           string                  `json:"id,omitempty"`
```

**Review by flbla — CHANGES_REQUESTED** [review]

File: docs/resources/config_security.md [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  cve_allowlist = ["CVE-456", "CVE-123"]
+  expires_at = "1701167767"
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `cve_allowlist` - (Optional) System allowlist. Vulnerabilities in this list will be ignored when pushing and pulling images. Should be in the format or `["CVE-123", "CVE-145"]` or `["CVE-123"]`
```
Comment (flbla) [inline docs/resources/config_security.md:16]:
```suggestion
* `cve_allowlist` - (Required) System allowlist. Vulnerabilities in this list will be ignored when pushing and pulling images. Should be in the format or `["CVE-123", "CVE-145"]` or `["CVE-123"]`
```

File: models/systemcve.go [hunk 10]
```diff
@@ -0,0 +1,17 @@
+package models
+var PathSystemCVEAllowList = "/system/CVEAllowlist"
+type SystemCveAllowListBodyPost struct {
+	ID           string                  `json:"id,omitempty"`
```
Comment (flbla) [inline models/systemcve.go:4]:
```suggestion
	ID           int                     `json:"id,omitempty"`
```

File: provider/resource_config_security.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+	resp, _, respCode, err := apiClient.SendRequest("GET", models.PathSystemCVEAllowList, nil, 200)
+	if respCode == 404 && err != nil {
+		d.SetId("")
+		return fmt.Errorf("resource not found %s", models.PathSystemCVEAllowList)
+	}
+
+	var jsonData models.SystemCveAllowListBodyPost
+	err = json.Unmarshal([]byte(resp), &jsonData)
+	if err != nil {
+		return fmt.Errorf("resource not found %s", d.Id())
```
Comment (flbla) [inline provider/resource_config_security.go:69]:
```suggestion
		return fmt.Errorf("resource not found %s", models.PathSystemCVEAllowList)
```

File: provider/resource_config_security.go [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+	expires_at, expires_at_true := d.GetOk("expires_at")
+
+	if expires_at_true {
+		d.Set("expires_at", expires_at)
+	}
+
+	d.SetId(models.PathSystemCVEAllowList)
+	d.Set("update_time", jsonData.UpdateTime)
+	d.Set("creation_time", jsonData.CreationTime)
```
Comment (flbla) [inline provider/resource_config_security.go:81]:
```suggestion
	// Convert the list of SystemCveAllowlistItems to a list of strings
	allowlistItems := make([]string, len(jsonData.Items))
	for i, item := range jsonData.Items {
		allowlistItems[i] = item.CveID
	}

	d.SetId(models.PathSystemCVEAllowList)
	d.Set("update_time", jsonData.UpdateTime)
	d.Set("expires_at", jsonData.ExpiresAt)
	d.Set("cve_allowlist", allowlistItems)
	d.Set("creation_time", jsonData.CreationTime)
```

File: provider/resource_config_security.go [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+	if err != nil {
+		return err
+	}
+
+	return resourceConfigSecurityRead(d, m)
+}
+
+func resourceConfigSecurityDelete(d *schema.ResourceData, m interface{}) error { // Harbor doesn't really have aby way to delete this resource yet...
+	apiClient := m.(*client.Client)
+	body := client.SystemCVEAllowListBody(d)
```
Comment (flbla) [inline provider/resource_config_security.go:100]:
```suggestion
func resourceConfigSecurityDelete(d *schema.ResourceData, m interface{}) error {
	apiClient := m.(*client.Client)
	d.Set("expires_at", nil)
	d.Set("cve_allowlist", []string{})
	body := client.SystemCVEAllowListBody(d)
```

File: provider/resource_config_security.go [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+	d.Set("cve_allowlist", []string{})
+	body := client.SystemCVEAllowListBody(d)
+
+	_, _, _, err := apiClient.SendRequest("PUT", models.PathSystemCVEAllowList, body, 200)
+	if err != nil {
+		return err
+	}
+
+	return resourceConfigSecurityRead(d, m)
+}
```
Comment (flbla) [inline provider/resource_config_security.go:111]:
```suggestion
func resourceConfigSecurityUpdate(d *schema.ResourceData, m interface{}) error {
	return resourceConfigSecurityCreate(d, m)
}

func resourceConfigSecurityDelete(d *schema.ResourceData, m interface{}) error {
	d.Set("expires_at", nil)
	d.Set("cve_allowlist", []string{})
	return resourceConfigSecurityCreate(d, m)
}
```

**Review by flbla — CHANGES_REQUESTED** [review]

**Pull request merged**